### DISCUSSION TITLE
Passing `parserOptions` to shareable config

### DISCUSSION THREAD

**Discussion by zaicevas** [eslint/eslint#15517] [Q&A]
Hey there. I am implementing a shareable config (let's call it `eslint-config-foo`) and I've faced a problem. Ideally, I want the following API in a project that is using shareable config:
```
// .eslintrc.js
module.exports = {
  extends: ['@foo', '@foo/eslint-config/typescript'],
  ...
}
```

However, in order to support type-aware rules, I have to pass:
```
      parserOptions: {
        tsconfigRootDir: __dirname,
        project: './tsconfig.json',
      },
```

Is it possible to pass this in `eslint-config-foo`, rather than in consumers? If no, is it possible to pass `__dirname` from the consumer project to `eslint-config-foo` so that `eslint-config-foo` can pass it? I found a hack to parse this in `eslint-config-foo` with:
```
const path = require('path')

const pathToEslint = path.dirname(require.main.filename)
const consumerProjectRoot = pathToEslint.split('node_modules')[0]
module.exports = {
      parserOptions: {
        tsconfigRootDir: consumerProjectRoot,
        project: `${consumerProjectRoot}/tsconfig.json`,
      },
      ...
}
```

~~However, it seems it VSCode linting doesn't work with the above solution.~~ (VSCode doesn't seem to work with `pnpm link` regardless of `parserOptions` hack). The simplest solution would be to expose even more presets, e.g. `@foo`, `@foo/eslint-config/typescript`, `@foo/eslint-config/typescript-tests` and then in projects have this:
```
// .eslintrc.js
module.exports = {
  overrides: [
  {
      files: ['*.ts', '*.tsx'],
      extends: ['@foo/eslint-config/typescript'],
      parserOptions: {
        tsconfigRootDir: consumerProjectRoot,
        project: `${consumerProjectRoot}/tsconfig.json`,
      },
  },
  {
      files: ['*.test.tsx', '*.test.tsx'],
      extends: ['@foo/eslint-config/typescript-tests'],
  },
  ],
  extends: ['@foo', '@foo/eslint-config/typescript'],
  ...
}
```

But this is much more ugly on consumer side, so I'd rather avoid it.

**Comment by nzakas** [comment]
Sorry, I’m not sure what you mean when you say “pass.” You can’t pass arguments into shareable configs. 

Shareable configs are designed to be extended. If you want someone to use your config with minimal setup, then it should contain all of the options that a consumer needs. If something needs to be dynamically calculated based on the project, then that needs to be in the project config.

**Reply by zaicevas** [reply]
Thanks for quick reply!

> Sorry, I’m not sure what you mean when you say “pass.” You can’t pass arguments into shareable configs.

Got it. By `pass` I meant "to have this option in eslint config".

Seems like a bigger issue that I am now facing is making VSCode ESLint extension to resolve `extends` in shareable ESLint config. If all of the `eslint-plugin-foo` (which uses `extends` to extend from airbnb and other configs) dependencies (`eslint-config-airbnb`, `eslint-config-airbnb-typescript`, `eslint`, `typescript`, etc.) are in `dependencies` in `package.json`, then VSCode is able to lint correctly. If all of the dependencies are peer dependencies, then VSCode fails to lint:

![image](https://user-images.githubusercontent.com/34719980/149487114-20c7e5d3-09ed-45c1-b9bd-574ad756a0e2.png)

Using `npx eslint`, it lints correctly in both cases (peer dependencies and `dependencies`). So, is it a standard practice for shareable configs to put everything, except plugins, into `dependencies` (`typescript`, `eslint`, other shareable configs) ?

**Reply by zaicevas** [reply]
Example of a shareable config:
```
// eslintrc.js eslint-config-foo
module.exports = {
  parser: '@babel/eslint-parser',
  extends: ['airbnb'],
  env: {
    browser: true,
    node: true,
    jest: true,
  },
  overrides: [
    {
      files: ['*.test.ts', '*.test.tsx', '*.test.js', '*.test.jsx'],
      plugins: ['testing-library'],
      rules: {
        'testing-library/await-async-query': 'error',
      },
    },
  ],
}
```

```
// package.json eslint-config-foo
  "dependencies": {
    "eslint": "^8.6.0",
    "eslint-config-airbnb": "^19.0.4",
    "typescript": "^4.5.4",
    "eslint-plugin-import": "^2.25.4",
    "eslint-plugin-jsx-a11y": "^6.5.1",
    "eslint-plugin-react": "^7.28.0",
    "eslint-plugin-react-hooks": "^4.3.0",
    "@babel/eslint-parser": "^7.16.5",
    "@babel/core": "^7.12.7",
    "eslint-config-typescript": "3.0.0"
  },
```

Consumer project:
```
// eslintrc.js
module.exports = {
  extends: ['foo']
}
```

**Reply by nzakas** [reply]
Most shareable configs use peer dependencies. Direct dependencies won’t resolve correctly
